Soomsil-USaint/Data/Client/RuSaint/GradeClient.swift

```diff
@@ -4,4 +4,253 @@
 //
 //  Created by 정지혁 on 1/9/25.
 //
+import CoreData
 
+import ComposableArchitecture
+import Rusaint
+
+struct GradeClient {
+    static let coreDataStack: CoreDataStack = .shared
+    
+    var fetchAllSemesterGrades: @Sendable () async throws -> [SemesterGrade]
+    var fetchGrades: @Sendable (_ year: Int, _ semester: SemesterType) async throws -> [ClassGrade]
+    
+    var getAllSemesterGrades: () async throws -> [CDSemester]
+    var getGrades: (_ year: Int, _ semester: String) async throws -> CDSemester?
+    var updateAllSemesterGrades: (_ rusaintSemesterGrades: [GradeSummary]) async throws -> Void
+    var updateGrades: (_ year: Int, _ semester: String, _ newLectures: [LectureDetail]) async throws -> Void
+    var updateGPA: (_ year: Int, _ semester: String, _ gpa: Float) async throws -> Void
+    var addGrades: (_ newSemester: GradeSummary) async throws -> Void
+    var deleteAllSemesterGrades: () async throws -> Void
+    var deleteGrades: (_ year: Int, _ semester: String) async throws -> Void
+}
+
+extension DependencyValues {
+    var gradeClient: GradeClient {
+        get { self[GradeClient.self] }
+        set { self[GradeClient.self] = newValue }
+    }
+}
+
+extension GradeClient: DependencyKey {
+    static var liveValue: GradeClient = {
+        @Dependency(\.studentClient) var studentClient: StudentClient
+        
+        return GradeClient(
+            fetchAllSemesterGrades: {
+                let session = try await studentClient.createSaintSession()
+                let response = try await CourseGradesApplicationBuilder()
+                    .build(session: session)
+                    .semesters(courseType: CourseType.bachelor)
+                return response
+            },
+            fetchGrades: { year, semester in
+                let session = try await studentClient.createSaintSession()
```

Comment:
```suggestion
        return GradeClient(
            fetchAllSemesterGrades: {
                @Dependency(\.studentClient) var studentClient: StudentClient
                let session = try await studentClient.createSaintSession()
                let response = try await CourseGradesApplicationBuilder()
                    .build(session: session)
                    .semesters(courseType: CourseType.bachelor)
                return response
            },
            fetchGrades: { year, semester in
                @Dependency(\.studentClient) var studentClient: StudentClient
                let session = try await studentClient.createSaintSession()
```

GradeClient를 만들면서 studentClient의 값을 한 번 가져오는 것에 대해서 우려점이 있습니다.

1. 아직까지는 StudentClient나 그 내부가 바뀔 가능성이 없기 때문에 크게 문제는 없을 것 같은데, 만약 어떤 이유로 값이 바뀔 경우, 참조 타입이 아니기 때문에 해당 변화를 제대로 반영할 수 없다는 것
2. 싱글톤 패턴과 달리 TCA에서 `@DependencyValue`를 사용하는 이유는 실행 시점에 의존성 주입을 받기 위함으로 알고 있어서, 현재 아키텍쳐에 적절한 지 의문이 있습니다.

그래서, 괜찮다면, 중복된 코드지만 호출되는 시점에 해당 값을 가져오는 것이 적절해보입니다!

Reply:
아니면 fetchAllSemesterGrades가 필요한 부분에서 UsaintSession을 파라미터로 넘겨 받는건 어떨까요?
그러면 실행시점에서 Session을 받아서 사용할 수 있을 것 같아서요!

Reply:
StudentClient 내부에 값을 저장하지 않고, 해당 함수가 호출될 때마다 키체인에서 값을 불러와 세션을 생성하는 로직이라 해당 부분에 대해서는 고려하지 않았는데 변경될 가능성을 생각한다면 호출되는 시점에 값을 가져오는 것이 좋을 것 같네요!

Reply:
UsaintSession을 파라미터로 받아온다면, 리듀서에서 `StudentClient`를 통해 Session을 받아와 파라미터를 넘겨주는 걸까요? 
제이든이 제안한 방법과 비교하여 어떠한 이점이 있을까요?! @LeeJoEun-01

Reply:
네! 제가 위에서 말씀들인건 리듀서에서 StudentClient를 통해 Session을 받아오는 방식입니다.
이점이 있다기 보다는 Client내에서 또 다른 Client의 정보를 가져오는 것이 옳바른 방식인지 의문이 들어서요!

Reply:
아하! 저도 Client내에서 또 다른 Client를 가져오는 것이 좋은 방법인지 고민이 있었는데, 의존을 추가하는 것보다 리듀서에서 넣어주는 방식이 좋을 것 같네요.

Reply:
리듀서에서 파라미터로 받아오는 방식이 `var fetchAllSemesterGrades: @Sendable (SaintSession) async throws -> [SemesterGrade]` 이런 방식을 말씀하시는 것이죠?

Reply:
넵, 맞습니다!